### ISSUE TITLE
DNS doesn't work when using unbound

### ISSUE THREAD

**Issue by schklom** [pi-hole/docker-pi-hole#812]
This is a: **Run Issue**  

## Details

When using https://gitlab.com/klutchell/unbound/, Pihole receives all DNS requests, but the reply never reaches the sender.

## Related Issues
- [x] I have searched this repository/Pi-hole forums for existing issues and pull requests that look similar 

## How to reproduce the issue 

1. Environment data
  * Operating System: **Debian Buster**
  * Hardware: RasPi 4B
  * Kernel Architecture:
```
$ uname -srvmpio
Linux 5.9.0-0.bpo.5-arm64 #1 SMP Debian 5.9.15-1~bpo10+1 (2020-12-31) aarch64 unknown unknown GNU/Linux
```
  * Docker Install Info and version: 
    - Software source: official rootless docker
    - Supplimentary Software: none, just docker commands
  * Hardware architecture: ARMv7, x64

2. docker-compose.yml contents, docker run shell command, or paste a screenshot of any UI based configuration of containers here
```
version: "3.8"

services:
  pihole:
    image: pihole/pihole
    container_name: pihole
    hostname: pihole
    ports:
       - "53:53/tcp"
       - "53:53/udp"
      - "8888:80/tcp"
    dns:
      - 127.0.0.1
      - 1.1.1.1
    environment:
      TZ: ${TZ}
      WEBPASSWORD_FILE: /run/secrets/pihole_password
      PIHOLE_DNS_: 172.80.2.2#5053
      SKIPGRAVITYONBOOT: 1
    secrets:
      - pihole_password
    networks:
      pihole-unbound:

  unbound:
    image: klutchell/unbound
    container_name: unbound
    networks:
      pihole-unbound:
        ipv4_address: 172.80.2.2

networks:
  pihole-unbound:
    ipam:
      config:
        - subnet: "172.80.2.0/24"
```
3. any additional info to help reproduce

Wherever I make `curl` or `nslookup` requests from, Pihole receives them, but answer never comes back.
The only way I'm connected on Internet now is because I use a regular 2ndary DNS in my router.

This happens when using unbound: specifying usual upstream DNS servers instead of `unbound` solves the problem, but I would rather avoid it.

I have checked `unbound` and the default configuration works fine:
```
$ docker exec unbound drill -p 5053 facebook.com @127.0.0.1
;; ->>HEADER<<- opcode: QUERY, rcode: NOERROR, id: 45643
;; flags: qr rd ra ; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; QUESTION SECTION:
;; facebook.com.        IN      A

;; ANSWER SECTION:
facebook.com.   300     IN      A       31.13.72.36

;; AUTHORITY SECTION:

;; ADDITIONAL SECTION:

;; Query time: 168 msec
;; SERVER: 127.0.0.1
;; WHEN: Thu Mar  4 17:29:15 2021
;; MSG SIZE  rcvd: 46
```
which leads me to think the problem comes from `Pihole`, but I don't know why.

Do you have any idea ? ^^

I will be very grateful to whoever can help :)

## These common fixes didn't work for my issue

- [x] I have tried removing/destroying my container, and re-creating a new container
- [x] I have tried fresh volume data by backing up and moving/removing the old volume data
- [ ] I have tried running the stock `docker run` example(s) in the readme (removing any customizations I added)
- [x] I have tried a newer or older version of Docker Pi-hole (depending what version the issue started in for me)
- [x] I have tried running without my volume data mounts to eliminate volumes as the cause

If the above debugging / fixes revealed any new information note it here.
Add any other debugging steps you've taken or theories on root cause that may help.

### EDIT
not using `SKIPGRAVITYONBOOT: 1` makes pihole unable to start
```
::: All config checks passed, cleared for startup ...
::: Enabling Query Logging
  [i] Enabling logging...
  [✓] Logging has been enabled!
 ::: Docker start setup complete
  [✗] DNS resolution is currently unavailable
```
I should point out (if it makes a difference) that I'm using Rootless Docker.
I should also say that I had the same issue when using regular Docker.

**Comment by dschaper** [comment]
https://gitlab.com/klutchell/unbound/-/issues/34

**Comment by schklom** [comment]
Thanks for the suggestion @dschaper :)
Unfortunately that setup breaks pihole as it fails to update gravity since DNS resolution doesn't work.
I even made sure to disable ufw
```
$ sudo ufw disable
Firewall stopped and disabled on system startup
```
but the logs don't change
```
$ docker logs pihole
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] 01-resolver-resolv: applying...
[fix-attrs.d] 01-resolver-resolv: exited 0.
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 20-start.sh: executing...
 ::: Starting docker specific checks & setup for docker pihole/pihole

  [i] Installing configs from /etc/.pihole...
  [i] Existing dnsmasq.conf found... it is not a Pi-hole file, leaving alone!
  [✓] Copying 01-pihole.conf to /etc/dnsmasq.d/01-pihole.conf
Setting DNS servers based on PIHOLE_DNS_ variable
Setting password: secretpassword
+ pihole -a -p secretpassword secretpassword
  [✓] New password set
DNSMasq binding to default interface: eth0
Added ENV to php:
                        "PHP_ERROR_LOG" => "/var/log/lighttpd/error.log",
                        "ServerIP" => "0.0.0.0",
                        "VIRTUAL_HOST" => "0.0.0.0",
Using IPv4 and IPv6
::: setup_blocklists now setting default blocklists up:
::: TIP: Use a docker volume for /etc/pihole/adlists.list if you want to customize for first boot
::: Blocklists (/etc/pihole/adlists.list) now set to:
https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
::: Testing pihole-FTL DNS: FTL started!
::: Testing lighttpd config: Syntax OK
::: All config checks passed, cleared for startup ...
::: Enabling Query Logging
  [i] Enabling logging...
  [✓] Logging has been enabled!
 ::: Docker start setup complete
  [✗] DNS resolution is currently unavailable
```

**Comment by edgd1er** [comment]
TL;DR unbound is only accepting requests in private ip ranges.

As I'm discovering unbound, these inputs might be wrong.

I tried the given yaml and I have the same behaviour as you do. Pihole requests are refused:

```
root@pihole:/# dig -p 5053 www.github.com @unbound

; <<>> DiG 9.11.5-P4-5.1+deb10u2-Debian <<>> -p 5053 www.free.fr @unbound
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: REFUSED, id: 35061
;; flags: qr rd ad; QUERY: 0, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; Query time: 0 msec
;; SERVER: 172.80.2.2#5053(172.80.2.2)
;; WHEN: Thu Mar 04 22:27:25 CET 2021
;; MSG SIZE  rcvd: 12

```
Resolution from the outside the container is ok:

```
/# nslookup -port=5053 www.github.com 127.0.0.1
Serveur :   localhost
Address:  127.0.0.1

Réponse ne faisant pas autorité :
Nom :    github.com
Address:  140.82.121.4
Aliases:  www.github.com
````

My guess is that unbound is filtering the [requesting client](https://gitlab.com/klutchell/unbound/-/blob/main/unbound.conf), hence the refused.

```
        access-control: 127.0.0.1/32 allow
	access-control: 192.168.0.0/16 allow
	access-control: 172.16.0.0/12 allow
	access-control: 10.0.0.0/8 allow
```
After changing DNS_PIHOLE from 172.80.2.2#5053 to PIHOLE_DNS_: 127.0.0.1#5053
```
nslookup -port=53 www.github.com 127.0.0.1
Serveur :   localhost
Address:  127.0.0.1

Réponse ne faisant pas autorité :
Nom :    github.com
Address:  140.82.121.4
Aliases:  www.github.com
``` 

I can think of three solutions:
- mount a unbound.conf that matches your IP range.
- use 127.0.0.1#5053 as PIHOLE_DNS as suggested in the README. (may have a performance impact as routes are not straight)
- change your subnet config for an [allowed private range](https://www.arin.net/reference/research/statistics/address_filters/) 

this is the output you may have, when the subnet is allowed:
```
 dig -p 5053 www.github.com @unbound

; <<>> DiG 9.11.5-P4-5.1+deb10u2-Debian <<>> -p 5053 www.github.com @unbound
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 33900
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;www.github.com.                        IN      A

;; ANSWER SECTION:
www.github.com.         3600    IN      CNAME   github.com.
github.com.             60      IN      A       140.82.121.3

;; Query time: 31 msec
;; SERVER: 172.16.2.2#5053(172.16.2.2)
;; WHEN: Thu Mar 04 23:03:40 CET 2021
;; MSG SIZE  rcvd: 73
``` 


If I may ask, what would be the reasons to add unbound to pihole, beside the DNSSEC ?

**Comment by schklom** [comment]
@edgd1er I cannot thank you enough. I'll make a donation to Pihole (I assume you're with them ?) for the amazing people they have, and the community they help for free :)

To answer you:
> mount a unbound.conf that matches your IP range

I haven't tried, because I don't know how to configure it and because I don't want to have yet another configuration file that I will spend hours to debug in a few years when something breaks ^^
> use 127.0.0.1#5053 as PIHOLE_DNS as suggested in the README. (may have a performance impact as routes are not straight)

That's actually what I tried next, but it didn't work for me.
> change your subnet config for an [allowed private range](https://www.arin.net/reference/research/statistics/address_filters/)

This worked perfectly.
I had to take time to manually define all my network subnets and `daemon.json` because I would get random subnet pool clashes
```
ERROR: Pool overlaps with other one on this address space
```
But now, it's working perfectly ! Thank you som much again ;)
> what would be the reasons to add unbound to pihole, beside the DNSSEC ?

The main benefit for me is to avoid telling Google or someone else what websites I visit for the only advantage of gaining 1-2 seconds per uncached request. I don't mind waiting that long when I go on a new website if the benefit is privacy.

**Comment by edgd1er** [comment]
@schklom ,

I'm glad I could help you. I'm just a simple, enthusiast,  pi-hole user, helping other users whenever I can. I was interested by the chaining of unbound to pi-hole. Not sure I will use it, but knowing it exists and can be installed is good enough for me at the moment.
